operator/builtin/input/aws/cloudwatch/cloudwatch.go

```diff
@@ -0,0 +1,322 @@
+		}
+	}
+
+	if c.logStreamNamePrefix != "" {
+		logStreamNamePrefixPtr := aws.String(c.logStreamNamePrefix)
+		return cloudwatchlogs.FilterLogEventsInput{
+			Limit:               limit,
+			LogGroupName:        logGroupNamePtr,
+			LogStreamNamePrefix: logStreamNamePrefixPtr,
+			StartTime:           aws.Int64(c.startTime),
+		}
+	}
+
+	if nextToken != "" {
+		nextTokenPtr := aws.String(nextToken)
+		return cloudwatchlogs.FilterLogEventsInput{
+			Limit:        limit,
+			LogGroupName: logGroupNamePtr,
+			StartTime:    aws.Int64(c.startTime),
+			NextToken:    nextTokenPtr,
+		}
+	}
+
+	return cloudwatchlogs.FilterLogEventsInput{
+		Limit:        limit,
+		LogGroupName: logGroupNamePtr,
+		StartTime:    aws.Int64(c.startTime),
+	}
+}
+
+// handleEvent is the handler for a AWS Cloudwatch Logs Filtered Event.
+func (c *CloudwatchInput) handleEvent(ctx context.Context, event *cloudwatchlogs.FilteredLogEvent) error {
+	// c.wg.Add(1)
+
+	e := make(map[string]interface{})
+	e["message"] = event.Message
+	e["event_id"] = event.EventId
+	e["log_stream_name"] = event.LogStreamName
+	e["timestamp"] = event.Timestamp
+	e["ingestion_time"] = event.IngestionTime
+
+	entry, err := c.NewEntry(nil)
+	if err != nil {
+		c.Error("Failed to create new entry from record", zap.Error(err))
+	}
+	// entry.Timestamp = time.Unix(0, *event.Timestamp*int64(time.Millisecond))
+	entry.Timestamp = FromUnixMilli(*event.Timestamp)
+	entry.Record = e
+	// Persist
+	if *event.IngestionTime > c.startTime {
+		c.persist.Set(c.logGroupName, Int64ToBytes(*event.IngestionTime))
+	}
+	// Write Entry
+	c.Write(ctx, entry)
+	// c.wg.Done()
+	return nil
+}
+
+func (c *CloudwatchInput) handleBatchedEvents(ctx context.Context, events []*cloudwatchlogs.FilteredLogEvent) error {
+	c.wg.Add(1)
+	defer c.wg.Done()
+
+	// Create an entry for each log in the batch.
+	wg := sync.WaitGroup{}
+	max := 10
+	gaurd := make(chan struct{}, max)
+	for i := 0; i < len(events); i++ {
+		e := events[i]
+		wg.Add(1)
+		gaurd <- struct{}{}
+		go func() {
+			defer func() {
+				wg.Done()
+				<-gaurd
+			}()
+			c.handleEvent(ctx, e)
+		}()
+	}
+	wg.Wait()
+	return nil
+}
+
+// Returns time.Now() as Unix Time in Milliseconds
+func currentTimeInUnixMilliseconds() int64 {
+	return time.Now().UnixNano() / int64(time.Millisecond)
+}
+
+// Helper function to persist start time
+func Int64ToBytes(i int64) []byte {
+	var buf = make([]byte, 8)
+	binary.BigEndian.PutUint64(buf, uint64(i))
+	return buf
+}
+
+// Helper function to get persisted start time
+func BytesToInt64(buf []byte) int64 {
+	var startTime int64
+	buffer := bytes.NewBuffer(buf)
+	binary.Read(buffer, binary.BigEndian, &startTime)
+	return startTime
+}
+
+// Helper function to convert Unix epoch time in milliseconds to go time
+func FromUnixMilli(ms int64) time.Time {
+	const millisInSecond = 1000
+	const nsInSecond = 1000000
+	return time.Unix(ms/int64(millisInSecond), (ms%int64(millisInSecond))*int64(nsInSecond))
+}
```

Comment:
I think these functions should not be exported as part of the public interface.

Reply:
Good catch.